specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+
+Basics
+---------
+
+#. A DataSet can store data of (reasonably) arbitrary types.
+#. A completed DataSet should be immutable; neither its metadata nor its results may be modified.
+
+Creation
+------------
+
+#. It should be possible to create a DataSet without knowing the final item count of the various values it stores. 
+   In particular, the number of loop iterations for a sweep should not be required to create the DataSet.
+#. The list of parameters in each result to be stored in a DataSet should be specified at creation time.
+   This includes the name, role (set-point or output), and type of each parameter.
+   Parameters may be marked as optional, in which case they are not required for each result.
+#. It should be possible to define a result parameter that is independent of any QCoDeSParameter or Instrument.
+#. A QCoDeS Parameter should provide sufficient information to define a result parameter.
+#. A DataSet should allow storage of relatively arbitrary metadata describing the run that generated the results and the parameters included in the results.
+    
+Writing
+----------
+
+#. It should be possible to add a single result or a sequence of results to an in-progress DataSet.
+#. A DataSet should maintain the order in which results were added.
+#. An in-progress DataSet may be marked as completed.
+
+Access
+---------
+
+#. Values in a DataSet should be easily accessible for plotting and analysis, even while the DataSet is in progress.
+   In particular, it should be possible to retrieve results in a NumPy-compatible form.
+#. It should be possible to define a cursor that specifies a location in a specific value set in a DataSet.
+   It should be possible to get a cursor that specifies the current end of the DataSet when the DataSet is "in progress".
+   It should be possible to read "new data" in a DataSet; that is, to read everything after a cursor.
+#. It should be possible to subscribe to change notifications from a DataSet.
+   It is acceptable if such subscriptions must be in-process until QCoDeS multiprocessing is redone.
+   Change notifications should include the results that were added to the DataSet that triggered the notification.
+
+Storage and Persistence
+-----------------------
+
+#. A DataSet object should allow writing to and reading from storage in a variety of formats.
+#. Users should be able to define new persistence formats.
+#. Users should be able to specify where a DataSet is written.
+
+Interface
+=========
+
+Creation
+--------
+
+ParamSpec
+~~~~~~~~~
+
+A ParamSpec object specifies a single parameter in a DataSet.
+
+ParamSpec(Parameter p, optional=)
+    Creates a parameter specification from a QCoDeS Parameter.
+    If optional is provided and is true, then the parameter is optional in each result.
+    
+ParamSpec(name, role, type, desc=, optional=)
+    Creates a parameter specification with the given name, role (‘I’ or ‘O’), and type. 
+    The type should be a NumPy dtype object.
```

Comment:
@giulioungaretti the thing I was thinking about is probably more similar to your idea than I may make it seem :). 

Let's see if I can explain it more concisely by an example: 

There is an ExperimentContainer (to be renamed as per @majacassidy 's arguments) that contains several DataSets, let's say 7. 
We label these datasets A, B, and C0, C1, C2, C3, C4. 

DataSet A is a simple dataset that is independent of all the others. 
DataSet B is an experiment containing 5 datapoints of some parameter y, corresponding to a specific setting for some parameter x. For each of those Datapoints a calibration was done before measuring parameter y.
DataSet s C0-C4 contain the DataSet corresponding to the calibration done for each point in DataSet B. 

I would propose that instead of just linking the DataSets relationally (as proposed by @giulioungaretti ) we go one step further, putting a the GUID of DataSets Cx as entries in DataSet B. 
This will achieve several goals. 
 - It is possible to relate datasets to specific points in Datasets (as in relational database) 
- DataSets are still flat, you can access DataSet C0-C4 directly 
- It is possible to infinitely nest or group DataSets (as I would like) 
- A viewer could then allow you to browse through nested files. 

This would then also make the notion of the ExperimentContainer a lot simpler, You can use it as a filesystem on steroids and it can contain any number of DataSets (or part of your DataSets) you want, it will only give a DataSet not available error if you happen to link to a DataSet that is not part of your container.